File: client/javascript/access_control_client.js
/**
 * Blockchain Access Control Client Library (JavaScript/TypeScript)
 * Client for interacting with the decentralized access control system
 */

const { ethers } = require('ethers');
const fs = require('fs');

class AccessControlClient {
  /**
   * Initialize the client
   * @param {string} providerUrl - Web3 provider URL (e.g., Infura endpoint)
   * @param {string} contractAddress - Deployed AccessControlManager contract address
   * @param {string} privateKey - Private key for signing transactions (optional for read-only)
   */
  constructor(providerUrl, contractAddress, privateKey = null) {
    this.provider = new ethers.providers.JsonRpcProvider(providerUrl);
    this.contractAddress = contractAddress;
    
    // Load contract ABI
    const contractABI = JSON.parse(
      fs.readFileSync('./abis/AccessControlManager.json', 'utf8')
    ).abi;
    
    if (privateKey) {
      this.wallet = new ethers.Wallet(privateKey, this.provider);
      this.contract = new ethers.Contract(
        contractAddress,
        contractABI,
        this.wallet
      );
      this.address = this.wallet.address;
    } else {
      this.contract = new ethers.Contract(
        contractAddress,
        contractABI,
        this.provider
      );
      this.wallet = null;
      this.address = null;
    }
  }

  /**
   * Create a new role
   * @param {string} roleName - Name of the role
   * @returns {Promise<{roleId: string, txHash: string}>}
   */
  async createRole(roleName) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const roleId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(roleName));
    const tx = await this.contract.createRole(roleId, roleName);
    const receipt = await tx.wait();
    
    return {
      roleId,
      txHash: receipt.transactionHash
    };
  }

  /**
   * Grant a role to a user
   * @param {string} roleName - Name of the role
   * @param {string} userAddress - Ethereum address of the user
   * @returns {Promise<string>} Transaction hash
   */
  async grantRole(roleName, userAddress) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const roleId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(roleName));
    const tx = await this.contract.grantRole(roleId, userAddress);
    const receipt = await tx.wait();
    
    return receipt.transactionHash;
  }

  /**
   * Revoke a role from a user
   * @param {string} roleName - Name of the role
   * @param {string} userAddress - Ethereum address of the user
   * @returns {Promise<string>} Transaction hash
   */
  async revokeRole(roleName, userAddress) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const roleId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(roleName));
    const tx = await this.contract.revokeRole(roleId, userAddress);
    const receipt = await tx.wait();
    
    return receipt.transactionHash;
  }

  /**
   * Add a permission to a role
   * @param {string} roleName - Name of the role
   * @param {string} resource - Resource identifier
   * @param {string} action - Action identifier
   * @returns {Promise<string>} Transaction hash
   */
  async addPermission(roleName, resource, action) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const roleId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(roleName));
    const permission = ethers.utils.keccak256(
      ethers.utils.toUtf8Bytes(`${resource}:${action}`)
    );
    
    const tx = await this.contract.addPermission(roleId, permission);
    const receipt = await tx.wait();
    
    return receipt.transactionHash;
  }

  /**
   * Check if a user has a specific role
   * @param {string} roleName - Name of the role
   * @param {string} userAddress - Ethereum address of the user
   * @returns {Promise<boolean>}
   */
  async hasRole(roleName, userAddress) {
    const roleId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(roleName));
    return await this.contract.hasRole(roleId, userAddress);
  }

  /**
   * Check if a user has permission to perform an action
   * @param {string} userAddress - Ethereum address of the user
   * @param {string} resource - Resource identifier
   * @param {string} action - Action identifier
   * @returns {Promise<boolean>}
   */
  async checkPermission(userAddress, resource, action) {
    const resourceBytes = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(resource));
    const actionBytes = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(action));
    
    return await this.contract.checkPermission(
      userAddress,
      resourceBytes,
      actionBytes
    );
  }

  /**
   * Get all roles assigned to a user
   * @param {string} userAddress - Ethereum address of the user
   * @returns {Promise<string[]>} Array of role IDs
   */
  async getUserRoles(userAddress) {
    const roles = await this.contract.getUserRoles(userAddress);
    return roles.map(r => r);
  }

  /**
   * Get all permissions for a role
   * @param {string} roleName - Name of the role
   * @returns {Promise<string[]>} Array of permission IDs
   */
  async getRolePermissions(roleName) {
    const roleId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(roleName));
    const permissions = await this.contract.getRolePermissions(roleId);
    return permissions.map(p => p);
  }

  /**
   * Get all roles in the system
   * @returns {Promise<string[]>} Array of role IDs
   */
  async getAllRoles() {
    const roles = await this.contract.getAllRoles();
    return roles.map(r => r);
  }

  /**
   * Get user information
   * @param {string} userAddress - Ethereum address of the user
   * @returns {Promise<Object>}
   */
  async getUserInfo(userAddress) {
    const info = await this.contract.getUserInfo(userAddress);
    return {
      isActive: info[0],
      addedAt: new Date(info[1].toNumber() * 1000),
      roleCount: info[2].toNumber()
    };
  }

  /**
   * Listen for role granted events
   * @param {Function} callback - Callback function(roleId, user, grantedBy)
   */
  onRoleGranted(callback) {
    this.contract.on('RoleGranted', (roleId, user, grantedBy, event) => {
      callback(roleId, user, grantedBy, event);
    });
  }

  /**
   * Listen for role revoked events
   * @param {Function} callback - Callback function(roleId, user, revokedBy)
   */
  onRoleRevoked(callback) {
    this.contract.on('RoleRevoked', (roleId, user, revokedBy, event) => {
      callback(roleId, user, revokedBy, event);
    });
  }

  /**
   * Pause the contract (emergency only)
   * @returns {Promise<string>} Transaction hash
   */
  async pauseContract() {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const tx = await this.contract.pause();
    const receipt = await tx.wait();
    return receipt.transactionHash;
  }

  /**
   * Unpause the contract
   * @returns {Promise<string>} Transaction hash
   */
  async unpauseContract() {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const tx = await this.contract.unpause();
    const receipt = await tx.wait();
    return receipt.transactionHash;
  }
}

class PolicyRegistryClient {
  /**
   * Initialize the policy registry client
   * @param {string} providerUrl - Web3 provider URL
   * @param {string} contractAddress - Deployed PolicyRegistry contract address
   * @param {string} privateKey - Private key for signing transactions
   */
  constructor(providerUrl, contractAddress, privateKey = null) {
    this.provider = new ethers.providers.JsonRpcProvider(providerUrl);
    this.contractAddress = contractAddress;
    
    const contractABI = JSON.parse(
      fs.readFileSync('./abis/PolicyRegistry.json', 'utf8')
    ).abi;
    
    if (privateKey) {
      this.wallet = new ethers.Wallet(privateKey, this.provider);
      this.contract = new ethers.Contract(
        contractAddress,
        contractABI,
        this.wallet
      );
    } else {
      this.contract = new ethers.Contract(
        contractAddress,
        contractABI,
        this.provider
      );
      this.wallet = null;
    }
  }

  /**
   * Create a new policy
   * @param {string} policyName - Name of the policy
   * @param {string} description - Policy description
   * @param {string} ipfsHash - IPFS hash of the detailed policy document
   * @returns {Promise<{policyId: string, txHash: string}>}
   */
  async createPolicy(policyName, description, ipfsHash) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const policyId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(policyName));
    const ipfsHashBytes = ethers.utils.arrayify(ipfsHash);
    
    const tx = await this.contract.createPolicy(
      policyId,
      policyName,
      description,
      ipfsHashBytes
    );
    const receipt = await tx.wait();
    
    return {
      policyId,
      txHash: receipt.transactionHash
    };
  }

  /**
   * Update an existing policy
   * @param {string} policyName - Name of the policy
   * @param {string} newIpfsHash - New IPFS hash
   * @param {string} changeDescription - Description of changes
   * @returns {Promise<string>} Transaction hash
   */
  async updatePolicy(policyName, newIpfsHash, changeDescription) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const policyId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(policyName));
    const ipfsHashBytes = ethers.utils.arrayify(newIpfsHash);
    
    const tx = await this.contract.updatePolicy(
      policyId,
      ipfsHashBytes,
      changeDescription
    );
    const receipt = await tx.wait();
    
    return receipt.transactionHash;
  }

  /**
   * Get policy information
   * @param {string} policyName - Name of the policy
   * @returns {Promise<Object>}
   */
  async getPolicy(policyName) {
    const policyId = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(policyName));
    const result = await this.contract.getPolicy(policyId);
    
    return {
      name: result[0],
      description: result[1],
      ipfsHash: ethers.utils.hexlify(result[2]),
      version: result[3].toNumber(),
      createdAt: new Date(result[4].toNumber() * 1000),
      updatedAt: new Date(result[5].toNumber() * 1000),
      createdBy: result[6],
      lastUpdatedBy: result[7],
      isActive: result[8]
    };
  }

  /**
   * Anchor a batch of logs on the blockchain
   * @param {string} batchHash - IPFS hash of the log batch
   * @param {number} logCount - Number of logs in the batch
   * @returns {Promise<string>} Transaction hash
   */
  async anchorLogBatch(batchHash, logCount) {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }
    
    const batchHashBytes = ethers.utils.arrayify(batchHash);
    
    const tx = await this.contract.anchorLogBatch(batchHashBytes, logCount);
    const receipt = await tx.wait();
    
    return receipt.transactionHash;
  }

  /**
   * Verify integrity of a log batch
   * @param {number} batchId - ID of the batch
   * @param {string} expectedHash - Expected IPFS hash
   * @returns {Promise<boolean>}
   */
  async verifyLogBatch(batchId, expectedHash) {
    const expectedHashBytes = ethers.utils.arrayify(expectedHash);
    return await this.contract.verifyLogBatch(batchId, expectedHashBytes);
  }

  /**
   * Listen for policy created events
   * @param {Function} callback - Callback function(policyId, name, ipfsHash, creator)
   */
  onPolicyCreated(callback) {
    this.contract.on('PolicyCreated', (policyId, name, ipfsHash, creator, event) => {
      callback(policyId, name, ipfsHash, creator, event);
    });
  }

  /**
   * Listen for policy updated events
   * @param {Function} callback - Callback function(policyId, newVersion, ipfsHash, updatedBy)
   */
  onPolicyUpdated(callback) {
    this.contract.on('PolicyUpdated', (policyId, newVersion, ipfsHash, updatedBy, event) => {
      callback(policyId, newVersion, ipfsHash, updatedBy, event);
    });
  }
}

// Example usage
async function example() {
  const PROVIDER_URL = 'https://sepolia.infura.io/v3/YOUR_INFURA_KEY';
  const ACCESS_CONTROL_ADDRESS = '0x...'; // Your deployed contract address
  const PRIVATE_KEY = '0x...'; // Your private key
  
  // Initialize client
  const client = new AccessControlClient(
    PROVIDER_URL,
    ACCESS_CONTROL_ADDRESS,
    PRIVATE_KEY
  );
  
  try {
    // Create a role
    const { roleId, txHash } = await client.createRole('Administrator');
    console.log(`Created role: ${roleId}`);
    console.log(`Transaction: ${txHash}`);
    
    // Grant role to user
    const userAddress = '0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb';
    const grantTxHash = await client.grantRole('Administrator', userAddress);
    console.log(`Granted role to ${userAddress}`);
    console.log(`Transaction: ${grantTxHash}`);
    
    // Check if user has role
    const hasRole = await client.hasRole('Administrator', userAddress);
    console.log(`User has Administrator role: ${hasRole}`);
    
    // Add permission
    const permTxHash = await client.addPermission('Administrator', 'database', 'write');
    console.log(`Added permission to role`);
    console.log(`Transaction: ${permTxHash}`);
    
    // Check permission
    const canWrite = await client.checkPermission(userAddress, 'database', 'write');
    console.log(`User can write to database: ${canWrite}`);
    
    // Listen for events
    client.onRoleGranted((roleId, user, grantedBy, event) => {
      console.log(`Role ${roleId} granted to ${user} by ${grantedBy}`);
    });
    
  } catch (error) {
    console.error('Error:', error);
  }
}

module.exports = {
  AccessControlClient,
  PolicyRegistryClient
};

File: client/python/access_control_client.py
"""
Blockchain Access Control Client Library
Python client for interacting with the decentralized access control system
"""

from web3 import Web3
from web3.middleware import geth_poa_middleware
from eth_account import Account
import json
import os
from typing import List, Dict, Optional, Tuple
from datetime import datetime
import hashlib

class AccessControlClient:
    """Client for interacting with AccessControlManager contract"""
    
    def __init__(self, provider_url: str, contract_address: str, private_key: Optional[str] = None):
        """
        Initialize the client
        
        Args:
            provider_url: Web3 provider URL (e.g., Infura endpoint)
            contract_address: Deployed AccessControlManager contract address
            private_key: Private key for signing transactions (optional for read-only)
        """
        self.w3 = Web3(Web3.HTTPProvider(provider_url))
        
        # Add PoA middleware for testnets like Sepolia
        self.w3.middleware_onion.inject(geth_poa_middleware, layer=0)
        
        if not self.w3.isConnected():
            raise ConnectionError("Failed to connect to Web3 provider")
        
        self.contract_address = Web3.toChecksumAddress(contract_address)
        
        # Load contract ABI
        with open('abis/AccessControlManager.json', 'r') as f:
            contract_abi = json.load(f)
        
        self.contract = self.w3.eth.contract(
            address=self.contract_address,
            abi=contract_abi['abi']
        )
        
        if private_key:
            self.account = Account.from_key(private_key)
            self.address = self.account.address
        else:
            self.account = None
            self.address = None
    
    def _send_transaction(self, function, *args, **kwargs):
        """Internal method to send transactions"""
        if not self.account:
            raise ValueError("Private key required for transactions")
        
        # Build transaction
        tx = function(*args, **kwargs).buildTransaction({
            'from': self.address,
            'nonce': self.w3.eth.getTransactionCount(self.address),
            'gas': 500000,
            'gasPrice': self.w3.eth.gas_price
        })
        
        # Sign transaction
        signed_tx = self.w3.eth.account.sign_transaction(tx, self.account.key)
        
        # Send transaction
        tx_hash = self.w3.eth.send_raw_transaction(signed_tx.rawTransaction)
        
        # Wait for receipt
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        
        return receipt
    
    def create_role(self, role_name: str) -> Tuple[str, str]:
        """
        Create a new role
        
        Args:
            role_name: Name of the role
            
        Returns:
            Tuple of (role_id, transaction_hash)
        """
        role_id = Web3.keccak(text=role_name)
        receipt = self._send_transaction(
            self.contract.functions.createRole,
            role_id,
            role_name
        )
        return (role_id.hex(), receipt['transactionHash'].hex())
    
    def grant_role(self, role_name: str, user_address: str) -> str:
        """
        Grant a role to a user
        
        Args:
            role_name: Name of the role
            user_address: Ethereum address of the user
            
        Returns:
            Transaction hash
        """
        role_id = Web3.keccak(text=role_name)
        user_address = Web3.toChecksumAddress(user_address)
        receipt = self._send_transaction(
            self.contract.functions.grantRole,
            role_id,
            user_address
        )
        return receipt['transactionHash'].hex()
    
    def revoke_role(self, role_name: str, user_address: str) -> str:
        """
        Revoke a role from a user
        
        Args:
            role_name: Name of the role
            user_address: Ethereum address of the user
            
        Returns:
            Transaction hash
        """
        role_id = Web3.keccak(text=role_name)
        user_address = Web3.toChecksumAddress(user_address)
        receipt = self._send_transaction(
            self.contract.functions.revokeRole,
            role_id,
            user_address
        )
        return receipt['transactionHash'].hex()
    
    def add_permission(self, role_name: str, resource: str, action: str) -> str:
        """
        Add a permission to a role
        
        Args:
            role_name: Name of the role
            resource: Resource identifier
            action: Action identifier
            
        Returns:
            Transaction hash
        """
        role_id = Web3.keccak(text=role_name)
        permission = Web3.keccak(text=f"{resource}:{action}")
        receipt = self._send_transaction(
            self.contract.functions.addPermission,
            role_id,
            permission
        )
        return receipt['transactionHash'].hex()
    
    def has_role(self, role_name: str, user_address: str) -> bool:
        """
        Check if a user has a specific role
        
        Args:
            role_name: Name of the role
            user_address: Ethereum address of the user
            
        Returns:
            True if user has the role, False otherwise
        """
        role_id = Web3.keccak(text=role_name)
        user_address = Web3.toChecksumAddress(user_address)
        return self.contract.functions.hasRole(role_id, user_address).call()
    
    def check_permission(self, user_address: str, resource: str, action: str) -> bool:
        """
        Check if a user has permission to perform an action
        
        Args:
            user_address: Ethereum address of the user
            resource: Resource identifier
            action: Action identifier
            
        Returns:
            True if user has permission, False otherwise
        """
        user_address = Web3.toChecksumAddress(user_address)
        resource_bytes = Web3.keccak(text=resource)
        action_bytes = Web3.keccak(text=action)
        return self.contract.functions.checkPermission(
            user_address,
            resource_bytes,
            action_bytes
        ).call()
    
    def get_user_roles(self, user_address: str) -> List[str]:
        """
        Get all roles assigned to a user
        
        Args:
            user_address: Ethereum address of the user
            
        Returns:
            List of role IDs (hex strings)
        """
        user_address = Web3.toChecksumAddress(user_address)
        roles = self.contract.functions.getUserRoles(user_address).call()
        return [r.hex() for r in roles]
    
    def get_role_permissions(self, role_name: str) -> List[str]:
        """
        Get all permissions for a role
        
        Args:
            role_name: Name of the role
            
        Returns:
            List of permission IDs (hex strings)
        """
        role_id = Web3.keccak(text=role_name)
        permissions = self.contract.functions.getRolePermissions(role_id).call()
        return [p.hex() for p in permissions]
    
    def get_all_roles(self) -> List[str]:
        """
        Get all roles in the system
        
        Returns:
            List of role IDs (hex strings)
        """
        roles = self.contract.functions.getAllRoles().call()
        return [r.hex() for r in roles]
    
    def pause_contract(self) -> str:
        """
        Pause the contract (emergency only, requires admin)
        
        Returns:
            Transaction hash
        """
        receipt = self._send_transaction(self.contract.functions.pause)
        return receipt['transactionHash'].hex()
    
    def unpause_contract(self) -> str:
        """
        Unpause the contract (requires admin)
        
        Returns:
            Transaction hash
        """
        receipt = self._send_transaction(self.contract.functions.unpause)
        return receipt['transactionHash'].hex()


class PolicyRegistryClient:
    """Client for interacting with PolicyRegistry contract"""
    
    def __init__(self, provider_url: str, contract_address: str, private_key: Optional[str] = None):
        """Initialize the policy registry client"""
        self.w3 = Web3(Web3.HTTPProvider(provider_url))
        self.w3.middleware_onion.inject(geth_poa_middleware, layer=0)
        
        if not self.w3.isConnected():
            raise ConnectionError("Failed to connect to Web3 provider")
        
        self.contract_address = Web3.toChecksumAddress(contract_address)
        
        with open('abis/PolicyRegistry.json', 'r') as f:
            contract_abi = json.load(f)
        
        self.contract = self.w3.eth.contract(
            address=self.contract_address,
            abi=contract_abi['abi']
        )
        
        if private_key:
            self.account = Account.from_key(private_key)
            self.address = self.account.address
        else:
            self.account = None
            self.address = None
    
    def _send_transaction(self, function, *args, **kwargs):
        """Internal method to send transactions"""
        if not self.account:
            raise ValueError("Private key required for transactions")
        
        tx = function(*args, **kwargs).buildTransaction({
            'from': self.address,
            'nonce': self.w3.eth.getTransactionCount(self.address),
            'gas': 500000,
            'gasPrice': self.w3.eth.gas_price
        })
        
        signed_tx = self.w3.eth.account.sign_transaction(tx, self.account.key)
        tx_hash = self.w3.eth.send_raw_transaction(signed_tx.rawTransaction)
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        
        return receipt
    
    def create_policy(
        self, 
        policy_name: str, 
        description: str, 
        ipfs_hash: str
    ) -> Tuple[str, str]:
        """
        Create a new policy
        
        Args:
            policy_name: Name of the policy
            description: Policy description
            ipfs_hash: IPFS hash of the detailed policy document
            
        Returns:
            Tuple of (policy_id, transaction_hash)
        """
        policy_id = Web3.keccak(text=policy_name)
        ipfs_hash_bytes = Web3.toBytes(hexstr=ipfs_hash)
        
        receipt = self._send_transaction(
            self.contract.functions.createPolicy,
            policy_id,
            policy_name,
            description,
            ipfs_hash_bytes
        )
        
        return (policy_id.hex(), receipt['transactionHash'].hex())
    
    def update_policy(
        self, 
        policy_name: str, 
        new_ipfs_hash: str, 
        change_description: str
    ) -> str:
        """
        Update an existing policy
        
        Args:
            policy_name: Name of the policy
            new_ipfs_hash: New IPFS hash
            change_description: Description of changes
            
        Returns:
            Transaction hash
        """
        policy_id = Web3.keccak(text=policy_name)
        ipfs_hash_bytes = Web3.toBytes(hexstr=new_ipfs_hash)
        
        receipt = self._send_transaction(
            self.contract.functions.updatePolicy,
            policy_id,
            ipfs_hash_bytes,
            change_description
        )
        
        return receipt['transactionHash'].hex()
    
    def get_policy(self, policy_name: str) -> Dict:
        """
        Get policy information
        
        Args:
            policy_name: Name of the policy
            
        Returns:
            Dictionary with policy information
        """
        policy_id = Web3.keccak(text=policy_name)
        result = self.contract.functions.getPolicy(policy_id).call()
        
        return {
            'name': result[0],
            'description': result[1],
            'ipfs_hash': result[2].hex(),
            'version': result[3],
            'created_at': datetime.fromtimestamp(result[4]),
            'updated_at': datetime.fromtimestamp(result[5]),
            'created_by': result[6],
            'last_updated_by': result[7],
            'is_active': result[8]
        }
    
    def anchor_log_batch(self, batch_hash: str, log_count: int) -> str:
        """
        Anchor a batch of logs on the blockchain
        
        Args:
            batch_hash: IPFS hash of the log batch
            log_count: Number of logs in the batch
            
        Returns:
            Transaction hash
        """
        batch_hash_bytes = Web3.toBytes(hexstr=batch_hash)
        
        receipt = self._send_transaction(
            self.contract.functions.anchorLogBatch,
            batch_hash_bytes,
            log_count
        )
        
        return receipt['transactionHash'].hex()
    
    def verify_log_batch(self, batch_id: int, expected_hash: str) -> bool:
        """
        Verify integrity of a log batch
        
        Args:
            batch_id: ID of the batch
            expected_hash: Expected IPFS hash
            
        Returns:
            True if batch is valid, False otherwise
        """
        expected_hash_bytes = Web3.toBytes(hexstr=expected_hash)
        return self.contract.functions.verifyLogBatch(
            batch_id,
            expected_hash_bytes
        ).call()


# Example usage
if __name__ == "__main__":
    # Configuration
    PROVIDER_URL = "https://sepolia.infura.io/v3/YOUR_INFURA_KEY"
    ACCESS_CONTROL_ADDRESS = "0x..."  # Your deployed contract address
    PRIVATE_KEY = "0x..."  # Your private key
    
    # Initialize client
    client = AccessControlClient(PROVIDER_URL, ACCESS_CONTROL_ADDRESS, PRIVATE_KEY)
    
    # Create a role
    role_id, tx_hash = client.create_role("Administrator")
    print(f"Created role: {role_id}")
    print(f"Transaction: {tx_hash}")
    
    # Grant role to user
    user_address = "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb"
    tx_hash = client.grant_role("Administrator", user_address)
    print(f"Granted role to {user_address}")
    print(f"Transaction: {tx_hash}")
    
    # Check if user has role
    has_role = client.has_role("Administrator", user_address)
    print(f"User has Administrator role: {has_role}")
    
    # Add permission
    tx_hash = client.add_permission("Administrator", "database", "write")
    print(f"Added permission to role")
    print(f"Transaction: {tx_hash}")
    
    # Check permission
    can_write = client.check_permission(user_address, "database", "write")
    print(f"User can write to database: {can_write}")
